Add generic settings to compliment site-specific settings (#8327)
wagtail/wagtail: docs/reference/contrib/settings.md

# Settings

> The `settings` module allows you to:

Author: I like the completeness of this, though am unsure about:

1. using the word "global" - recent feedback was to rename this work to "generic settings" instead of "global settings", as "global" may confuse things if we're working with a multilingual setup.
2. removing the two bullet points that currently follow this line; I think it's useful to clearly call out that there are two separate workflows.

Thoughts?

accessed in code as well as in templates.

## Installation

Add `wagtail.contrib.settings` to your `INSTALLED_APPS`:

```python
INSTALLED_APPS += [
    'wagtail.contrib.settings',
]
```

**Note:** If you are using `settings` within templates, you will also need to
update your `TEMPLATES` settings (discussed later in this page).

## Defining settings

Create a model that inherits from either:

* `BaseGenericSetting` (for generic settings across all sites)
* `BaseSiteSetting` (for site-specific settings)

and register it using the `register_setting` decorator:

```python
from django.db import models
from wagtail.contrib.settings.models import (
    BaseGenericSetting,
    BaseSiteSetting,
    register_setting,
)

@register_setting
class GenericSocialMediaSettings(BaseGenericSetting):
    facebook = models.URLField()

@register_setting
class SiteSpecificSocialMediaSettings(BaseSiteSetting):
    facebook = models.URLField()
```

Links to your settings will appear in the Wagtail admin 'Settings' menu.

## Edit handlers

Settings use edit handlers much like the rest of Wagtail.
Add a `panels` setting to your model defining all the edit handlers required:

```python
@register_setting

> class GenericImportantPages(BaseGenericSetting):

Reviewer: Generally, I don't think users should/would be prefixing these models with ``Generic``/``SiteSpecific``.

To make this slightly more realistic (and avoid establishing an unnecessary naming convention!), could we use two separate examples, one for each type?

For example: ``SocialMediaSettings(BaseGenericSetting)`` and ``ImportantPages(BaseSiteSetting)``

Author: I'm not sure there is a reliable real-world example we can include here that always makes sense. For instance, I need both generic and site-specific social media settings on a project I'm working on now. 

Let's be as explicitly clear as possible within the docs, and then it's up to the developer to decide on the model naming for their project.

    donate_page = models.ForeignKey(
        'wagtailcore.Page', null=True, on_delete=models.SET_NULL, related_name='+'
    )
    sign_up_page = models.ForeignKey(
        'wagtailcore.Page', null=True, on_delete=models.SET_NULL, related_name='+'
    )

    panels = [
        FieldPanel('donate_page'),
        FieldPanel('sign_up_page'),
    ]

@register_setting
class SiteSpecificImportantPages(BaseSiteSetting):
    donate_page = models.ForeignKey(
        'wagtailcore.Page', null=True, on_delete=models.SET_NULL, related_name='+'
    )
    sign_up_page = models.ForeignKey(
        'wagtailcore.Page', null=True, on_delete=models.SET_NULL, related_name='+'
    )

    panels = [
        FieldPanel('donate_page'),
        FieldPanel('sign_up_page'),
    ]
```

> You can also customize the editor handlers [like you would do for Page model](/advanced_topics/customisation/page_editing_interface.html#customising-the-tabbed-interface) with a custom `edit_handler` attribute:

Reviewer: ```suggestion
You can also customise the edit handlers [like you would do for `Page` model](/advanced_topics/customisation/page_editing_interface.html#customising-the-tabbed-interface) with a custom `edit_handler` attribute:
```

```python
from wagtail.admin.panels import TabbedInterface, ObjectList

@register_setting
class MySettings(BaseGenericSetting):
    # ...
    first_tab_panels = [
        FieldPanel('field_1'),
    ]
    second_tab_panels = [
        FieldPanel('field_2'),
    ]

    edit_handler = TabbedInterface([
        ObjectList(first_tab_panels, heading='First tab'),
        ObjectList(second_tab_panels, heading='Second tab'),
    ])
```

## Appearance

You can change the label used in the menu by changing the

> `django.db.models.Options.verbose_name` of your model.

Reviewer: Spelling out the full path to ``verbose_name`` feels a little verbose to me!

```suggestion
`verbose_name` of your model.
```

Author: Agreed - a lot of the copy on this page is lifted straight from the current `.rst` version with minimal amends.

You can add an icon to the menu by passing an `icon` argument to the
`register_setting` decorator:

```python
@register_setting(icon='placeholder')
class GenericSocialMediaSettings(BaseGenericSetting):
    ...
    class Meta:
        verbose_name = "Social media settings for all sites"

@register_setting(icon='placeholder')
class SiteSpecificSocialMediaSettings(BaseSiteSetting):
    ...
    class Meta:
        verbose_name = "Site-specific social media settings"
```

For a list of all available icons, please see the [styleguide](/contributing//styleguide.html).

## Using the settings

Settings can be used in both Python code and in templates.

### Using in Python

#### Generic settings

If you require access to a generic setting in a view, the

> `wagtail.contrib.settings.models.BaseGenericSetting.get_or_create` method allows

Reviewer: I feel like we could drop the module path here too

```suggestion
`BaseGenericSetting.get_or_create` method allows
```

Author: I find the full path helpful, though happy to take your lead for consistency here!

you to retrieve the generic settings:

```python
def view(request):
    social_media_settings = GenericSocialMediaSettings.get_or_create()
    ...
```

#### Site-specific settings

If you require access to a site-specific setting in a view, the

> `wagtail.contrib.settings.models.BaseSiteSetting.for_request` method allows

Reviewer: ```suggestion
`BaseSiteSetting.for_request` method allows
```

you to retrieve the site-specific settings for the current request:

```python
def view(request):
    social_media_settings = SiteSpecificSocialMediaSettings.for_request(request)
    ...
```

> In places where the request is unavailable, but you know the Site you wish to

Reviewer: ```suggestion
In places where the request is unavailable, but you know the `Site` you wish to
```

context. If you want to assign to a different context variable instead, use
`{% get_settings as other_variable_name %}`:

```html+django
{% load wagtailsettings_tags %}
{% get_settings as wagtail_settings %}
{{ wagtail_settings.app_label.GenericSocialMediaSettings.facebook }}
{{ wagtail_settings.app_label.SiteSpecificSocialMediaSettings.facebook }}
```

### Using in Jinja2 templates

Add `wagtail.contrib.settings.jinja2tags.settings` extension to your
Jinja2 settings:

```python
TEMPLATES = [
    ...

    {
        'BACKEND': 'django.template.backends.jinja2.Jinja2',
        'APP_DIRS': True,
        'OPTIONS': {
            'extensions': [
                ...

                'wagtail.contrib.settings.jinja2tags.settings',
            ],
        },
    }
]
```

Then access the settings through the `settings()` template function:

```html+jinja
{{ settings("app_label.GenericSocialMediaSettings").facebook }}
{{ settings("app_label.SiteSpecificSocialMediaSettings").facebook }}
```

**Note:** Replace `app_label` with the label of the app containing your
settings model.

If there is no `request` available in the template at all, you can use the

> settings for the default Wagtail site instead:

Reviewer: ```suggestion
settings for the default site instead:
```

{{ settings("app_label.SiteSpecificSocialMediaSettings", use_default_site=True).facebook }}
```

**Note:** You can not reliably get the correct settings instance for the
current site from this template tag if the request object is not available.
This is only relevant for multisite instances of Wagtail.

You can store the settings instance in a variable to save some typing,
if you have to use multiple values from one model:

```html+jinja
{% with generic_social_settings=settings("app_label.GenericSocialMediaSettings") %}
    Follow us on Twitter at @{{ generic_social_settings.facebook }},
    or Instagram at @{{ generic_social_settings.instagram }}.
{% endwith %}

{% with site_social_settings=settings("app_label.SiteSpecificSocialMediaSettings") %}
    Follow us on Twitter at @{{ site_social_settings.facebook }},
    or Instagram at @{{ site_social_settings.instagram }}.
{% endwith %}
```

Or, alternately, using the `set` tag:

```html+jinja
{% set generic_social_settings=settings("app_label.GenericSocialMediaSettings") %}
{% set site_social_settings=settings("app_label.SiteSpecificSocialMediaSettings") %}
```

## Utilising `select_related` to improve efficiency

> For models with foreign key relationships to other objects (e.g. pages),

Reviewer: ```suggestion
For models with foreign key relationships to other objects (for example, pages),
```

Author: We use `e.g.` elsewhere - what's the reasoning/thinking behind changing this one?

Author: TIL! Thanks for the references :)
